src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -1275,10 +1275,46 @@ fn traverse_and_share(
     } else {
         unit.kind
     };
+
+    let mut profile = unit.profile.clone();
+
+    // If this is a build dependency, and it's not shared with runtime dependencies, we can weaken
+    // its debuginfo level to optimize build times.
+    if unit.kind.is_host() && profile.debuginfo.is_deferred() {
```

Comment:
A build dep with artifact dependencies should keep debuginfo I think as it may end up embedded in the main program.

Reply:
If a build script needs to pre-process an artifact to for example generate bindings using wasm-bindgen, it has to be specified as build dependency, right?

Reply:
Sounds pretty valid and common to me. Thank you!

Then I'd suggest not nuking debuginfo from any kind of artifact dep, since we don't know users' intents.

Reply:
Yes agreed, here I was trying to find the `is_for_host` that `UnitFor` has and the `Profile` sees to set defaults and so on. 

But since the `UnitDep` is available when we recurse, I've pushed a commit to use `UnitFor::is_for_host` instead of `CompileKind::is_host`. That should be better, right ?

(I've also pushed a commit to not optimize debuginfo for artifact dependencies. I'm not sure what kind of test we should add to ensure that though, any thoughts ?)